### DISCUSSION TITLE
Browser extension

### DISCUSSION THREAD

**Discussion by erengy** [erengy/taiga#1084] [Ideas]
Taiga's streaming media detection feature relies on UI automation, where Taiga (via [Anisthesia](https://github.com/erengy/anisthesia)) acts as a screen reader in order to read the address bar and the tabs of a web browser. While this method works fine for the most part, we either cannot access the DOM or doing so dramatically decreases the performance. It also partially relies on the web browser's language setting to be in English. Some services such as Hulu and Netflix, and more recently Crunchyroll (#994) and Funimation (#1003), do not display the required information in their page title, which makes detecting the currently playing media technically impossible.

This brings us to another idea: Using a browser extension/addon that works together with Taiga. Basically, this extension should detect known streaming sites and extract the relevant information (anime title, season number, episode number, episode title, etc.), then send it to Taiga using [native messaging](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging) or a similar method (#129, #1045).

From what I remember, this wasn't feasible back when I first developed Taiga's streaming media detection feature in 2011. I still don't like the idea of requiring users to install an additional piece of software in order to enable a certain feature. But, with the recent developments, this might be a good time to reevaluate.

**Comment by snickler** [comment]
I believe this may be the best way to go. I was attempting to find ways to read the title from Crunchyroll with their updated nonsense and came to the same conclusion. If it requires extra hooks to read from the DOM etc, that's going to add a noticeable hit on performance. An extension would be the right way to go. Something that could communicate back to Taiga over a port would be pretty nice.

**Reply by alan1420** [reply]
Maybe something like RPC or WebSocket

**Comment by Californ1a** [comment]
It could potentially be a standalone browser extension, if you aren't playing any local media. Of course, you'd need the desktop app for monitoring local media players, but for web players, a [similar extension](https://github.com/web-scrobbler/web-scrobbler/) already exists for music scrobbling a ton of different sites to last.fm, meaning they've already got all the main work done for finding titles and metadata (such as play/pause state) on various web players. It's also written in such a way that it's very simple for someone to add a new "connector" for a different site. It would probably be possible to fork that, replace the last.fm login/api with AL, AP, etc. (rewrite the settings page), delete most the music-specific connectors, and add a couple connectors for anime sites. However, they are still using the old v2 manifest, so it may be better just to take that project as inspiration for how it could work, and the structure/layout with independent connectors, and start from scratch using the new v3 manifest.

As for non-standalone, communicating back to the desktop app, there is a built-in [`runtime.sendNativeMessage()`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendNativeMessage) function for browser extensions to communicate with native apps. As for replying back to the extension from the desktop app, there's a couple examples of that [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging#app_side).

**Reply by Bartek200219** [reply]
> It could potentially be a standalone browser extension

I think that taiga's big advantage is library is option to manage library fast and easy and I don't think that standalone extension would give that feel, besides option like that already exists [MALSync](https://github.com/MALSync/MALSync/)

**Reply by Californ1a** [reply]
Sure, MALSync exists as a standalone option, but I'm mostly thinking like if you have Taiga desktop installed on one PC but not another, you could still be able to use the browser extension independently on the other PC, and let the browser sync take care of copying stuff back to the first PC when you return to it, so it syncs back to the desktop app. Just having the option to be able to use it both with and without, and then sync to the desktop app if it is installed, rather than *requiring* both.
